### PR TITLE
NF: Refactor Ease button constants

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
We currently have two set of constants representing exactly the same things, the four buttons the user can click. This lead to small typing problem when going from one type to the other. I suggest using a single set of constants, the ones from Consts, so that we follow upstream.

Once this is done, I'd even suggest going to an enumeration, and only using actual ints when we really need to (i.e. for logging, and to take keyboard input)

### CODE REVIEW

**Comment by mikehardy** [comment]
Fails `./gradlew :AnkiDroid:lintRelease`

```
* What went wrong:

Execution failed for task ':AnkiDroid:lintRelease'.

> Lint found errors in the project; aborting build.

  

  Fix the issues identified by lint, or add the following to your build script to proceed with errors:

  ...

  android {

      lintOptions {

          abortOnError false

      }

  }

  ...

  

  Errors found:

  

  /home/travis/build/ankidroid/Anki-Android/AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:1349: Error: Must be one of: Consts.BUTTON_ONE, Consts.BUTTON_TWO, Consts.BUTTON_THREE, Consts.BUTTON_FOUR [WrongConstant]

                      return 0;

                             ~

  /home/travis/build/ankidroid/Anki-Android/AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:1354: Error: Must be one of: Consts.BUTTON_ONE, Consts.BUTTON_TWO, Consts.BUTTON_THREE, Consts.BUTTON_FOUR [WrongConstant]

              return 0;
```

**Review by mikehardy — CHANGES_REQUESTED** [review]
I like the idea - some shaping up to fix lintRelease and it could work

**Comment by Arthur-Milchior** [comment]
0 is not a button. It's closer to "null". So I'll directly add an enum and make it @Nullable

**Comment by Arthur-Milchior** [comment]
Rebased on top of https://github.com/ankidroid/Anki-Android/pull/7044 because I wanted to add annotations about what is and is not nullable, and it's easier to consider all at once.

Buttons are an enum. As a string, it's transformed into a number so that it can even be used directly in sql. This allow to even more explicitly write everywhere where a button is required. There is a method which allow to transform a number into a button, used as little as possible.

This resolve the problem of having 0, which represents "no button" which was treated the same way as actual buttons. Now 0 is transformed into null, and @nullable/@NonNull allow to check what kind of buttons we want

**Review by mikehardy — APPROVED** [review]
Seems reasonable. It's a large enough refactor that I'm willing to hold up the release cut just a bit to get this in first so we aren't straddling a refactor across the branches immediately. @david-allison-1 what do you think? Seems like the core of it is the enum change, the rest is derivative.

**Comment by david-allison-1** [comment]
Some of the naming is non-camelCase

Could you add a test to ensure that enum.toString() results in the correct integer (for SQL string generation).

Could we wait until the `@NonNull` PR dependency is merged, there's too much noise in this PR to appropriately review it.

**Comment by mikehardy** [comment]
I just reviewed commit by commit to de-noise it. Good call on the test and casing though

**Comment by david-allison-1** [comment]
Okay, reviewing commit-by-commit.

Main concern: does https://github.com/ankidroid/Anki-Android/pull/7041/commits/74a262b2f8d7b373656d72a1dc49a19ae7d1e8b7 compile (previously failed due to the 0 constant, but was that fixed)? If it doesn't, it'll cause sporadic `git bisect` failures

**Comment by david-allison-1** [comment]
https://github.com/ankidroid/Anki-Android/pull/7041/commits/7cac0f4685ada339772bc863dcf5a0ec94bceded : I'd prefer just returning an `Optional<Integer>` here - it's a far-reaching change, and I don't really see the benefit of the enum as we're breaking the enum abstraction all over the place to access and compare its internals.

@mikehardy @Arthur-Milchior thoughts? I'm aware this will be controversial and it's not a strong desire.

**Comment by mikehardy** [comment]
not terribly controversial but if it is not an easy "100% agree, no problem" for everyone then I'm mentally pushing this until after release-2.13 is cut / de-prioritizing it, so for me this will sit for a little while. Which is fine, I'll mention! No worries

**Comment by Arthur-Milchior** [comment]
Unit test added

> [7cac0f4](https://github.com/ankidroid/Anki-Android/commit/7cac0f4685ada339772bc863dcf5a0ec94bceded) : I'd prefer just returning an `Optional<Integer>` here - it's a far-reaching change, and I don't really see the benefit of the enum as we're breaking the enum abstraction all over the place to access and compare its internals.

I'm surprised by this statement. I juts counted, there are 105 occurrence of BUTTON_TYPE and 394 BUTTON_NUMBER. There are 9 calls to the extra methods 1 call to `fromInt`, 4 `getValue`, 1 `buttonsAtMostThis` and 3 `greaterThan`. Most of the time, methods simply uses switch or equality. So even if there is a few extra to enum, I would not state it's used all over the place.

It's not a pure enum, I'll admit it. But in Java, I find quite natural to represents buttons as a class. No method seems forced to me.  After all, once you know a card should have 3 buttons, getting the array with buttons 1 2 and 3 makes sens. Being able to check whether the user pressed something strictly greater than "again" makes sens also.

I would state that it adds a little bit of clarity, at the very least because it ensures that BUTTON_* is used everywhere we expect a button/answer; and also because the few methods added to the class have an intuitive meaning.
In particular, it is always slightly complex for me to recall whether buttons are 0 based or 1 based. So if the method tells the number of button is 3, whether the buttons should be [1, 2, 3], [0, 1, 2, 3] or [0, 1, 2]... so I find this version less complex to read

Of course, you may disagree, but that was my idea when doing this PR and I still prefer this version. Even if an optional over an integer may solve the initial problem too.

**Comment by Arthur-Milchior** [comment]
Conflict resolved

File: AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java [hunk 1]
```diff
@@ -374,10 +375,10 @@ public Cursor query(@NonNull Uri uri, String[] projection, String selection, Str
                     if (currentCard == null) {
                         break;
                     }
-                    int buttonCount = col.getSched().answerButtons(currentCard);
+                    @NonNull BUTTON_TYPE buttonCount = col.getSched().greatestAnswerButton(currentCard);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java:378]: answerButtons() is the count of answer buttons. "Greatest answer button" has less meaning to me, and we move further away from Anki Desktop.

File: AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java [hunk 2]
```diff
@@ -374,10 +375,10 @@ public Cursor query(@NonNull Uri uri, String[] projection, String selection, Str
...
                     JSONArray buttonTexts = new JSONArray();
-                    for (int i = 0; i < buttonCount; i++) {
-                        buttonTexts.put(col.getSched().nextIvlStr(mContext, currentCard, i + 1));
+                    for (BUTTON_TYPE button : buttonCount.buttonsAtMostThis()) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java:380]: I feel this looping construct is harder to reason about

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
After going through the code, I think the enum should be removed unless there's a good reason for it to stay - it feels that the code is larger, moving further from libAnki, more likely to contain bugs in the future (null/toString) and harder to reason about, and I don't see the benefits.

This is a far-reaching change into the scheduler/reviewer, and can break in a lot of places. I don't think the reward justifies the risks involved.

Consolidating the constants is an easy and good move, but the enum makes me hesitate.

**Comment by mikehardy** [comment]
failed unit tests in CI

**Pull request closed without merging**